File: scripts/export-data.py
#!/usr/bin/env python3
"""
Split Database Export Tool (Python Version)

Exports expense data for external analysis with pandas/numpy support.

USAGE:
    python scripts/export-data.py --format csv --view expenses
    python scripts/export-data.py --format parquet --view settlements
    python scripts/export-data.py --format all --output ./exports/

REQUIREMENTS:
    pip install psycopg2-binary python-dotenv pandas pyarrow openpyxl

FORMATS: csv, json, parquet, excel, pickle, sql-insert
VIEWS: expenses, settlements, categories, events, users, groups, trends, all
"""

import psycopg2
import pandas as pd
import json
import os
import sys
import argparse
from datetime import datetime
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables
load_dotenv('.env.local')

DATABASE_URL = os.getenv('DATABASE_URL')
if not DATABASE_URL:
    print("❌ DATABASE_URL not found in .env.local")
    sys.exit(1)

VIEWS = {
    'expenses': 'expense_summary_for_analysis',
    'settlements': 'user_settlement_summary_for_analysis',
    'categories': 'category_spending_for_analysis',
    'events': 'event_summary_for_analysis',
    'users': 'user_spending_profile_for_analysis',
    'groups': 'group_summary_for_analysis',
    'trends': 'daily_spending_trend_for_analysis',
}


def get_connection():
    """Create database connection"""
    try:
        conn = psycopg2.connect(DATABASE_URL)
        return conn
    except psycopg2.Error as e:
        print(f"❌ Connection failed: {e}")
        sys.exit(1)


def fetch_view_data(view_name):
    """Fetch data from a view into a pandas DataFrame"""
    conn = get_connection()
    try:
        print(f"🔍 Fetching data from {view_name}...")
        df = pd.read_sql_query(f"SELECT * FROM {view_name};", conn)
        print(f"   Found {len(df)} rows, {len(df.columns)} columns")
        return df
    except Exception as e:
        print(f"❌ Query failed: {e}")
        sys.exit(1)
    finally:
        conn.close()


def export_csv(df, output_path):
    """Export to CSV"""
    df.to_csv(output_path, index=False)
    print(f"✅ Exported to CSV: {output_path}")


def export_json(df, output_path):
    """Export to JSON"""
    df.to_json(output_path, orient='records', indent=2, date_format='iso')
    print(f"✅ Exported to JSON: {output_path}")


def export_parquet(df, output_path):
    """Export to Parquet (efficient for large datasets)"""
    try:
        df.to_parquet(output_path, engine='pyarrow')
        print(f"✅ Exported to Parquet: {output_path}")
    except ImportError:
        print("❌ Parquet export requires: pip install pyarrow")
        sys.exit(1)


def export_excel(df, output_path):
    """Export to Excel"""
    try:
        df.to_excel(output_path, index=False, sheet_name='Data')
        print(f"✅ Exported to Excel: {output_path}")
    except ImportError:
        print("❌ Excel export requires: pip install openpyxl")
        sys.exit(1)


def export_pickle(df, output_path):
    """Export to Pickle (Python native format)"""
    df.to_pickle(output_path)
    print(f"✅ Exported to Pickle: {output_path}")


def export_sql_insert(df, output_path, table_name='export_data'):
    """Generate SQL INSERT statements"""
    lines = []
    for _, row in df.iterrows():
        columns = ', '.join(df.columns)
        values = ', '.join([
            'NULL' if pd.isna(v) else f"'{str(v).replace(chr(39), chr(39)*2)}'" if isinstance(v, str) else str(v)
            for v in row
        ])
        lines.append(f"INSERT INTO {table_name} ({columns}) VALUES ({values});")
    
    with open(output_path, 'w') as f:
        f.write('\n'.join(lines))
    print(f"✅ Exported SQL inserts: {output_path}")


def get_file_size_str(path):
    """Get human-readable file size"""
    size_bytes = os.path.getsize(path)
    for unit in ['B', 'KB', 'MB', 'GB']:
        if size_bytes < 1024:
            return f"{size_bytes:.2f} {unit}"
        size_bytes /= 1024
    return f"{size_bytes:.2f} TB"


def export_view(view_key, output_format):
    """Export a single view"""
    if view_key not in VIEWS:
        print(f"❌ Unknown view: {view_key}")
        print(f"Available views: {', '.join(VIEWS.keys())}")
        sys.exit(1)

    view_name = VIEWS[view_key]
    date_str = datetime.now().strftime('%Y-%m-%d')
    output_path = f"export_{view_key}_{date_str}.{output_format}"
    
    print(f"📊 Exporting data from Split database\n")
    print(f"📋 View: {view_key}")
    print(f"📁 Output: {output_path}")
    print(f"📄 Format: {output_format}\n")

    df = fetch_view_data(view_name)
    
    if df.empty:
        print("⚠️  No data to export")
        return

    # Export based on format
    exporters = {
        'csv': export_csv,
        'json': export_json,
        'parquet': export_parquet,
        'excel': export_excel,
        'pickle': export_pickle,
        'sql-insert': lambda df, path: export_sql_insert(df, path),
    }

    if output_format not in exporters:
        print(f"❌ Unsupported format: {output_format}")
        print(f"Supported formats: {', '.join(exporters.keys())}")
        sys.exit(1)

    exporters[output_format](df, output_path)
    
    print(f"   Rows: {len(df)}")
    print(f"   Columns: {len(df.columns)}")
    print(f"   File size: {get_file_size_str(output_path)}\n")
    print(f"✅ Export complete!")


def export_all_views(output_format, output_dir='./exports'):
    """Export all views to separate files"""
    Path(output_dir).mkdir(exist_ok=True)
    
    print(f"📊 Exporting all views to {output_dir}\n")
    
    exported_files = []
    for view_key in VIEWS.keys():
        view_name = VIEWS[view_key]
        date_str = datetime.now().strftime('%Y-%m-%d')
        output_path = os.path.join(output_dir, f"export_{view_key}_{date_str}.{output_format}")
        
        print(f"🔄 Exporting {view_key}...")
        df = fetch_view_data(view_name)
        
        if not df.empty:
            exporters = {
                'csv': export_csv,
                'json': export_json,
                'parquet': export_parquet,
                'excel': export_excel,
                'pickle': export_pickle,
                'sql-insert': lambda df, path: export_sql_insert(df, path),
            }
            
            if output_format in exporters:
                exporters[output_format](df, output_path)
                exported_files.append((view_key, output_path, len(df)))
        else:
            print(f"   ⚠️  No data in {view_key}")
    
    print(f"\n✅ Exported {len(exported_files)} files:")
    for view_key, path, rows in exported_files:
        size = get_file_size_str(path)
        print(f"   - {view_key}: {rows} rows ({size})")


def main():
    parser = argparse.ArgumentParser(
        description='Export Split database views for external analysis'
    )
    parser.add_argument(
        '--view',
        default='expenses',
        help=f'View to export: {", ".join(VIEWS.keys())}, all'
    )
    parser.add_argument(
        '--format',
        default='csv',
        help='Export format: csv, json, parquet, excel, pickle, sql-insert'
    )
    parser.add_argument(
        '--output',
        help='Output file path (optional, auto-generated if not provided)'
    )
    parser.add_argument(
        '--output-dir',
        default='./exports',
        help='Output directory for --view all'
    )

    args = parser.parse_args()

    if args.view == 'all':
        export_all_views(args.format, args.output_dir)
    else:
        export_view(args.view, args.format)


if __name__ == '__main__':
    main()

File: scripts/export-data.js
#!/usr/bin/env node
/**
 * Split Database Export Tool
 * 
 * Exports expense data for external analysis tools (Tableau, Power BI, Excel, R, Python, etc.)
 * 
 * USAGE:
 *   node scripts/export-data.js --format csv --view expenses
 *   node scripts/export-data.js --format json --view settlements --output settlements.json
 *   node scripts/export-data.js --format csv --view all
 * 
 * FORMATS: csv, json, tsv, sql-insert
 * VIEWS: expenses, settlements, categories, events, users, groups, trends, all
 */

import pkg from 'pg';
const { Client } = pkg;
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';
import { createReadStream, createWriteStream } from 'fs';

dotenv.config({ path: '.env.local' });

const args = process.argv.slice(2);
const format = args.find(a => a.startsWith('--format'))?.split('=')[1] || 'csv';
const viewName = args.find(a => a.startsWith('--view'))?.split('=')[1] || 'expenses';
const output = args.find(a => a.startsWith('--output'))?.split('=')[1] || `export_${viewName}_${new Date().toISOString().split('T')[0]}.${format}`;

const views = {
  expenses: 'expense_summary_for_analysis',
  settlements: 'expense_splits_for_analysis',
  payers: 'user_payer_summary_for_analysis',
  participants: 'user_participant_summary_for_analysis',
  categories: 'category_spending_for_analysis',
  events: 'events_summary_for_analysis',
  groups: 'group_spending_for_analysis',
  trends: 'daily_spending_trend_for_analysis',
};

// ⚠️  IMPORTANT: All views include UNIQUE usernames for unambiguous identification
// Username fields (username) are unique in the database and safe for data merges
// Display name fields (first_name, last_name) are kept for UI readability but NOT unique
// Always use username as the primary key for integration with external tools

const client = new Client({
  connectionString: process.env.DATABASE_URL,
});

async function exportData() {
  try {
    await client.connect();
    console.log(`📊 Exporting data from Split database\n`);

    // Handle "all" view
    if (viewName === 'all') {
      console.log('🔄 Exporting all views...');
      const allExports = [];
      for (const [name, view] of Object.entries(views)) {
        const data = await fetchViewData(view);
        const fileName = `export_${name}_${new Date().toISOString().split('T')[0]}.${format}`;
        await writeFile(fileName, data, format);
        allExports.push(fileName);
      }
      console.log(`\n✅ Exported ${allExports.length} files:`);
      allExports.forEach(f => console.log(`   - ${f}`));
    } else {
      // Single view
      const viewName_trimmed = viewName.trim();
      const viewQuery = views[viewName_trimmed];
      
      if (!viewQuery) {
        console.error(`❌ Unknown view: ${viewName_trimmed}`);
        console.error(`Available views: ${Object.keys(views).join(', ')}`);
        process.exit(1);
      }

      console.log(`📋 View: ${viewName_trimmed}`);
      console.log(`📁 Output: ${output}`);
      console.log(`📄 Format: ${format}\n`);

      const data = await fetchViewData(viewQuery);
      const result = await writeFile(output, data, format);
      
      console.log(`✅ Export complete!`);
      console.log(`   Rows: ${data.length}`);
      console.log(`   File size: ${(result / 1024).toFixed(2)} KB`);
    }

  } catch (error) {
    console.error('❌ Export failed:', error.message);
    process.exit(1);
  } finally {
    await client.end();
  }
}

async function fetchViewData(viewName) {
  console.log(`🔍 Fetching data from ${viewName}...`);
  const result = await client.query(`SELECT * FROM ${viewName};`);
  console.log(`   Found ${result.rows.length} rows`);
  return result.rows;
}

async function writeFile(fileName, data, format) {
  if (data.length === 0) {
    console.warn('⚠️  No data to export');
    return 0;
  }

  let content;
  switch (format) {
    case 'csv':
      content = convertToCSV(data);
      break;
    case 'json':
      content = JSON.stringify(data, null, 2);
      break;
    case 'tsv':
      content = convertToCSV(data, '\t');
      break;
    case 'sql-insert':
      content = convertToSQL(data);
      break;
    default:
      throw new Error(`Unsupported format: ${format}`);
  }

  fs.writeFileSync(fileName, content);
  console.log(`✅ Saved to: ${fileName}`);
  
  return content.length;
}

function convertToCSV(rows, delimiter = ',') {
  if (rows.length === 0) return '';

  // Get headers from first row
  const headers = Object.keys(rows[0]);
  const headerRow = headers.map(h => `"${h}"`).join(delimiter);

  // Convert rows
  const csvRows = rows.map(row => {
    return headers.map(header => {
      const value = row[header];
      
      // Handle null/undefined
      if (value === null || value === undefined) {
        return '';
      }
      
      // Handle booleans
      if (typeof value === 'boolean') {
        return value ? 'TRUE' : 'FALSE';
      }
      
      // Handle strings with special characters
      if (typeof value === 'string') {
        if (value.includes(delimiter) || value.includes('"') || value.includes('\n')) {
          return `"${value.replace(/"/g, '""')}"`;
        }
        return `"${value}"`;
      }
      
      // Numbers and dates
      return String(value);
    }).join(delimiter);
  });

  return [headerRow, ...csvRows].join('\n');
}

function convertToSQL(rows) {
  if (rows.length === 0) return '';

  // This is a simplified version - for production, use proper SQL escaping
  const firstRow = rows[0];
  const columns = Object.keys(firstRow);
  const tableName = 'export_data';

  const statements = rows.map(row => {
    const values = columns.map(col => {
      const value = row[col];
      if (value === null || value === undefined) return 'NULL';
      if (typeof value === 'string') return `'${value.replace(/'/g, "''")}'`;
      if (typeof value === 'boolean') return value ? 'TRUE' : 'FALSE';
      return value;
    });
    
    return `INSERT INTO ${tableName} (${columns.join(', ')}) VALUES (${values.join(', ')});`;
  });

  return statements.join('\n');
}

// Run
exportData();
